### PR TITLE
Coordinator pattern implementation

### PR DESCRIPTION
**Pull request by IvanKulundzic**
Implemented the coordinator pattern, introducing the Coordinator protocol, main AppCoordinator, and login flow LoginCoordinator.

### CODE REVIEW

File: QuizApp/App/Common/SceneDelegate.swift [hunk 1]
```diff
@@ -3,6 +3,9 @@ import UIKit
 class SceneDelegate: UIResponder, UIWindowSceneDelegate {
 
     var window: UIWindow?
+    /// It is safe to force unwrap a coordinator here, as we want the app to crash
+    /// if something is not setup correctly.
```
Comment (istaudacher) [inline QuizApp/App/Common/SceneDelegate.swift:7]: This is ok.

**Review by istaudacher — CHANGES_REQUESTED** [review]
This app is simple and it will probably have just one NavCon. Let's use just one Coordinator for now.

File: QuizApp/App/Protocols/CoordinatorProtocol.swift [hunk 2]
```diff
@@ -0,0 +1,7 @@
+import UIKit
+
+protocol CoordinatorProtocol {
+
+    @discardableResult func start() -> UIViewController
```
Comment (istaudacher) [inline QuizApp/App/Protocols/CoordinatorProtocol.swift]: Coordinator should have functions that start with `show` and describe what screen should be presented.

File: QuizApp/App/Common/Coordinator.swift [hunk 3]
```diff
@@ -0,0 +1,24 @@
+import UIKit
+
+final class Coordinator: CoordinatorProtocol {
+
+    private var navigationController: UINavigationController?
```
Comment (istaudacher) [inline QuizApp/App/Common/Coordinator.swift]: Initiate NavCon in SceneDelegate and assign it to window. Pass it to Coordinator via init.

**Review by istaudacher — CHANGES_REQUESTED** [review]

File: QuizApp/App/Protocols/CoordinatorProtocol.swift (hunk 2 above)
Comment (IvanKulundzic) [reply QuizApp/App/Protocols/CoordinatorProtocol.swift]: Reworked it.

File: QuizApp/App/Protocols/CoordinatorProtocol.swift [hunk 4]
```diff
@@ -0,0 +1,7 @@
+import UIKit
+
+protocol CoordinatorProtocol {
+
+    @discardableResult func showLogin() -> UIViewController
```
Comment (istaudacher) [inline QuizApp/App/Protocols/CoordinatorProtocol.swift]: Coordinator functions should not return anything.

**Review by istaudacher — CHANGES_REQUESTED** [review]

File: QuizApp/App/Protocols/CoordinatorProtocol.swift (hunk 4 above)
Comment (IvanKulundzic) [reply QuizApp/App/Protocols/CoordinatorProtocol.swift]: Refactored.

File: QuizApp/App/Common/Coordinator.swift [hunk 5]
```diff
@@ -0,0 +1,25 @@
+import UIKit
+
+final class Coordinator: CoordinatorProtocol {
+
+    var navigationController: UINavigationController
```
Comment (istaudacher) [inline QuizApp/App/Common/Coordinator.swift]: This can be private let.

File: QuizApp/App/Common/Coordinator.swift [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+    }
+
+}
+
+// MARK: - Private methods
+private extension Coordinator {
+
+    func startLoginFlow() {
+        let loginViewController = LoginViewController()
+        navigationController.pushViewController(loginViewController, animated: true)
```
Comment (istaudacher) [inline QuizApp/App/Common/Coordinator.swift]: What if some screen is in the app and you want to navigate to the Login page? Is push the correct way to do it?

**Review by istaudacher — CHANGES_REQUESTED** [review]

File: QuizApp/App/Common/Coordinator.swift (hunk 5 above)
Comment (IvanKulundzic) [reply QuizApp/App/Common/Coordinator.swift]: Great, missed that one.

File: QuizApp/App/Common/Coordinator.swift (hunk 6 above)
Comment (IvanKulundzic) [reply QuizApp/App/Common/Coordinator.swift]: Agreed. I've changed it to use the `setViewControllers(_:animated:)` method. That way, in the example you mentioned, we would reset the navigation stack, without the need to pop view controllers.

**Review by istaudacher — APPROVED** [review]

**Pull request merged**